dvc/remote/local/__init__.py

```diff
@@ -581,3 +586,52 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if not self.changed(
+            path_info, {self.PARAM_CHECKSUM: checksum}
+        ) and self._is_same_link_as_cache(path_info):
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
+            return False
+        return True
+
+    def _get_cache_type(self, path_info):
+        if self.cache_type_confirmed:
+            return self.cache_types[0]
+
+        workspace_file = path_info.with_name("." + uuid())
+        test_cache_file = self.path_info / ".cache_type_test_file"
+        if not self.exists(test_cache_file):
```

Comment:
Just wanted to point out that this is racey, as cache might be shared and some other dvc instance might create test_cache_file after your `if not exists` check, but before open(). It doesn't really matter in this case, as the worst that could happen is that the file contents are going to be overwritten once, which is totally fine in this scenario.

Reply:
Probably in the future we should think of mechanism that would allow few instances to access same cache.